### PR TITLE
Remove Hyper-V as a hard dependency

### PR DESCRIPTION
**Pull request by aserdean**
Remove Hyper-V as a hard dependency

This patch includes an updated version of the OVS installer on Windows.

This patch also provides updates to the ansible playbooks to accommodate
for a non Hyper-V enabled host.

This patch also includes fixes and updates to the ansible playbooks.

Signed-off-by: Alin Gabriel Serdean <aserdean@ovn.org>
Signed-off-by: Alin-Gheorghe Balutoiu <alinbalutoiu@gmail.com>
Co-authored-by: Alin-Gheorghe Balutoiu <alinbalutoiu@gmail.com>

### CODE REVIEW

File: contrib/roles/linux/ovn-kubernetes/vars/ubuntu.yml [hunk 1]
```diff
@@ -2,9 +2,9 @@
 ovn_kubernetes_info:
   install_path: /usr/bin
   # ovn-kubernetes build info
-  git_url: https://github.com/openvswitch/ovn-kubernetes
+  git_url:  https://github.com/aserdean/ovn-kubernetes
   build_path: /tmp
-  branch: master
+  branch: no_hyperv
```
Comment (alinbalutoiu) [inline contrib/roles/linux/ovn-kubernetes/vars/ubuntu.yml]: This is just for testing, should be changed back to the original repo and master branch for final PR.

File: contrib/roles/windows/docker/tasks/requirements-16299.yaml [hunk 2]
```diff
@@ -8,9 +8,9 @@
   register: features_installed
   with_items:
     - Containers
-    - Hyper-V # TODO
-    - RSAT-Hyper-V-Tools # TODO
-    - Hyper-V-PowerShell # TODO
+    #- Hyper-V # TODO
+    #- RSAT-Hyper-V-Tools # TODO
+    #- Hyper-V-PowerShell # TODO
```
Comment (alinbalutoiu) [inline contrib/roles/windows/docker/tasks/requirements-16299.yaml]: Just remove those lines instead of commenting them.

File: contrib/roles/windows/docker/tasks/requirements-16299.yaml [hunk 3]
```diff
@@ -8,9 +8,9 @@
...
 
 - name: Windows 1709 | Expect reboot_required to false
```
Comment (alinbalutoiu) [inline contrib/roles/windows/docker/tasks/requirements-16299.yaml]: Rename the filename to ``requirements.yaml`` and put ``Windows Server`` instead of ``Windows 1709`` as those requirements will be for all versions of Windows.

File: contrib/roles/windows/kubernetes/tasks/setup_ovs_docker.yml [hunk 4]
```diff
@@ -12,42 +12,28 @@
       if (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole] "Administrator")) { Start-Process powershell.exe "-NoProfile -ExecutionPolicy Bypass -File `"$PSCommandPath`"" -Verb RunAs; exit }
       $RegROPath = "HKCU:\SOFTWARE\Microsoft\Windows\CurrentVersion\Run"
       Remove-ItemProperty $RegROPath "OVN_Init_Node" -ErrorAction SilentlyContinue
-
-      # Start-BitsTransfer https://raw.githubusercontent.com/Microsoft/SDN/master/Kubernetes/windows/hns.psm1
-      # Import-Module ./hns.psm1
+      #TODO figure why they don't get automatically included
+      Import-Module "C:\Windows\System32\WindowsPowerShell\v1.0\Modules\HostNetworkingService\HostNetworkingService.psm1"
+      Get-HNSNetwork
```
Comment (alinbalutoiu) [inline contrib/roles/windows/kubernetes/tasks/setup_ovs_docker.yml]: Is this line required?

File: contrib/roles/windows/kubernetes/tasks/setup_ovs_docker.yml [hunk 5]
```diff
@@ -12,42 +12,28 @@
...
       Restart-Service ovs-vswitchd
       # Clone the MAC addr
-      # TODO: this does not work if the mac ends in 99
-      Set-NetAdapter -Name "{{ interface_name }}" -MacAddress {{ interface_mac_addr[:-2] }}99 -Confirm:$false
-      Set-NetAdapter -Name br-ex -MacAddress {{ interface_mac_addr }} -Confirm:$false
-      Enable-NetAdapter br-ex
-      # Remove-NetIPAddress -ifAlias br-ex -Confirm:$false
-      # New-NetIPAddress -ifAlias br-ex -IPAddress {{ host_internal_ip }} -DefaultGateway {{ interface_default_gateway }} -PrefixLength {{ interface_prefix_length }}
-      # Set-DnsClientServerAddress -InterfaceAlias br-ex -ServerAddresses ("8.8.8.8")
+      # TODO: Carry over MTU to the management interface
```
Comment (alinbalutoiu) [inline contrib/roles/windows/kubernetes/tasks/setup_ovs_docker.yml]: This should be implemented in order to have this working on clouds that have MTU requirements (such as GCE).

File: go-controller/pkg/ovn/management-port.go [hunk 6]
```diff
@@ -19,7 +19,7 @@ const (
 func configureManagementPortWindows(clusterSubnet, clusterServicesSubnet,
 	routerIP, interfaceName, interfaceIP string) error {
 	// Up the interface.
-	args := []string{"Enable-NetAdapter", fmt.Sprintf("%s", interfaceName)}
+	args := []string{"Enable-NetAdapter", "-IncludeHidden", fmt.Sprintf("%s", interfaceName)}
```
Comment (alinbalutoiu) [inline go-controller/pkg/ovn/management-port.go]: Can you put the changes of the go files in a separate commit?

File: go-controller/pkg/util/ovs.go [hunk 7]
```diff
@@ -14,7 +14,7 @@ import (
 )
 
 const (
-	ovsCommandTimeout = 5
+	ovsCommandTimeout = 15
```
Comment (alinbalutoiu) [inline go-controller/pkg/util/ovs.go]: I think a comment would be good here explaining why the timeout has been increased that much.

**Comment by alinbalutoiu** [comment]
Added some comments to the PR.
I tested on Windows Server 1803 (buildnumber 17134 and all the updates installed) and encountered the following issues:
- it looks like HNS doesn't report that the extension is enabled even though it is actually enabled and running
```
@{Id=583cc151-73ec-4a6a-8b47-578297ad7623; IsEnabled=False; Name=Cloudbase Open vSwitch Extension},
```
this makes the call to Get-OVSEnabledHNSNetworks to fail since it won't find any HNS network that has the extension enabled. Changed the definition of Get-OVSEnabledHNSNetworks to return the HNS network without checking if extension is enabled allowed it to work. Note: it seems that not even restart fixes the issue and it still shows extension is disabled after restart.

- can you also change this check? https://github.com/openvswitch/ovn-kubernetes/blob/master/contrib/roles/windows/kubernetes/tasks/main.yml#L41
otherwise it will always run setup_ovs_docker on every playbook run. You should check if the bridge is created and skip the task, or even ``when: transparent_inspect.rc != 0`` should work fine as well

File: contrib/roles/windows/docker/tasks/requirements-16299.yaml (hunk 2 above)
Comment (aserdean) [reply contrib/roles/windows/docker/tasks/requirements-16299.yaml]: Hmmm. Should we bother we Hyper-V features via some variable?

File: contrib/roles/linux/ovn-kubernetes/vars/ubuntu.yml (hunk 1 above)
Comment (aserdean) [reply contrib/roles/linux/ovn-kubernetes/vars/ubuntu.yml]: Yup. Everything is very `alpha` here :)

File: contrib/roles/windows/docker/tasks/requirements-16299.yaml (hunk 3 above)
Comment (aserdean) [reply contrib/roles/windows/docker/tasks/requirements-16299.yaml]: Sure

File: contrib/roles/windows/kubernetes/tasks/setup_ovs_docker.yml (hunk 5 above)
Comment (aserdean) [reply contrib/roles/windows/kubernetes/tasks/setup_ovs_docker.yml]: Yup. Should I split this in a different PR?

File: go-controller/pkg/ovn/management-port.go (hunk 6 above)
Comment (aserdean) [reply go-controller/pkg/ovn/management-port.go]: Ofc

File: go-controller/pkg/util/ovs.go (hunk 7 above)
Comment (aserdean) [reply go-controller/pkg/util/ovs.go]: Sure.

File: go-controller/pkg/ovn/management-port.go (hunk 6 above)
Comment (aserdean) [reply go-controller/pkg/ovn/management-port.go]: https://github.com/openvswitch/ovn-kubernetes/pull/372
sent

File: contrib/roles/windows/kubernetes/tasks/setup_ovs_docker.yml (hunk 4 above)
Comment (aserdean) [reply contrib/roles/windows/kubernetes/tasks/setup_ovs_docker.yml]: Nope. I will remove it in the update

**Comment by aserdean** [comment]
@rajatchopra , @alinbalutoiu I updated the PR and tested on 1709 and also 1803. Can you please test it out?

**Comment by rajatchopra** [comment]
After a marathon session, Alin helped me get a cluster up and running with this PR. Thanks @aserdean .
This PR lgtm. Ready for merge, pending the branch name change.

The other issue of finding the default interface for docker network setup will dealt with separately.

File: contrib/roles/windows/kubernetes/tasks/main.yml [hunk 8]
```diff
@@ -38,7 +38,7 @@
 # and it's safer to have everything inside a powershell script to run.
 - name: Kubernetes | OVS Docker network setup
   include_tasks: ./setup_ovs_docker.yml
-  when: interface_name != "br-ex"
+  when: transparent_inspect.rc != 0
```
Comment (alinbalutoiu) [inline contrib/roles/windows/kubernetes/tasks/main.yml]: This check should be improved. I think the following should work:
```
- name: Kubernetes | Check if system-id is set
  win_shell: ovs-vsctl get Open_vSwitch . external_ids:system-id 
  register: ovs_system_id
  ignore_errors: true
  changed_when: false

- name: Kubernetes | OVS Docker network setup
  include_tasks: ./setup_ovs_docker.yml
  when: ovs_system_id.rc != 0
```

File: contrib/roles/windows/kubernetes/tasks/run_minion_init.yml [hunk 9]
```diff
@@ -19,6 +19,36 @@
...
+      win_service:
+        name: ovn-kubernetes-node
+        state: absent
+      changed_when: false
+  when: ovnkube_service.exists
+
+- name: Kubelet | Create ovn-kubernetes-node service
+  win_service:
+    name: ovn-kubernetes-node
+    path: '"{{ install_info.install_path }}\\servicewrapper.exe" ovn-kubernetes-node {{ install_info.install_path }}\\ovnkube.exe --k8s-kubeconfig={{ install_info.install_path }}\\kubeconfig.yaml --k8s-apiserver http://{{ kubernetes_info.MASTER_IP }}:8080 --init-node {{ ansible_hostname|lower }} --k8s-token {{TOKEN}} --nb-address "tcp://{{ kubernetes_info.MASTER_IP }}:6641" --sb-address "tcp://{{ kubernetes_info.MASTER_IP }}:6642" --cluster-subnet {{ kubernetes_info.CLUSTER_SUBNET }} --cni-conf-dir="{{ install_info.install_path }}/cni" --cni-plugin "ovn-k8s-cni-overlay.exe"'
```
Comment (alinbalutoiu) [inline contrib/roles/windows/kubernetes/tasks/run_minion_init.yml]: Please add the following argument ``--encap-ip={{ host_internal_ip }}`` to ovnkube.

File: contrib/roles/windows/kubernetes/tasks/run_minion_init.yml [hunk 10]
```diff
@@ -27,9 +57,8 @@
     line: |
       $RegROPath = "HKCU:\SOFTWARE\Microsoft\Windows\CurrentVersion\Run"
       Remove-ItemProperty $RegROPath "OVN_Init_Node" -ErrorAction SilentlyContinue
-      # TODO: Remove the sleep for the next release of OVS
-      sleep 30
-      {{ install_info.install_path }}\\ovnkube.exe --k8s-kubeconfig={{ install_info.install_path }}\\kubeconfig.yaml --k8s-apiserver http://{{ kubernetes_info.MASTER_IP }}:8080 --init-node {{ ansible_hostname|lower }} --k8s-token {{TOKEN}} --nb-address "tcp://{{ kubernetes_info.MASTER_IP }}:6641" --sb-address "tcp://{{ kubernetes_info.MASTER_IP }}:6642" --cluster-subnet {{ kubernetes_info.CLUSTER_SUBNET }} --cni-conf-dir="{{ install_info.install_path }}/cni" --cni-plugin "ovn-k8s-cni-overlay.exe"
+      Start-Service ovn-kubernetes-node
+      sleep 15
       ovs-vsctl set open_vswitch . external_ids:ovn-encap-ip="{{ host_public_ip }}"
```
Comment (alinbalutoiu) [inline contrib/roles/windows/kubernetes/tasks/run_minion_init.yml]: The sleep and setting ovn-encap-ip can be removed if ``--encap-ip={{ host_internal_ip }}`` is passed to ovnkube.

File: contrib/roles/windows/kubernetes/tasks/main.yml (hunk 8 above)
Comment (aserdean) [reply contrib/roles/windows/kubernetes/tasks/main.yml]: It still won't be sufficient to be honest.
I'll try to add some powershell commandlets to the OVS installer to improve the network creation/check/removal and send out a different PR for it.

File: contrib/roles/windows/kubernetes/tasks/run_minion_init.yml (hunk 9 above)
Comment (aserdean) [reply contrib/roles/windows/kubernetes/tasks/run_minion_init.yml]: I don't really like this approach because the adapter may change(i.e. DHCP server) . We need to improve the IP detection in the end IMO.
I will add it as a workaround and open an issue for the IP detection.

File: contrib/roles/windows/kubernetes/tasks/run_minion_init.yml (hunk 10 above)
Comment (aserdean) [reply contrib/roles/windows/kubernetes/tasks/run_minion_init.yml]: Same as above. I will add it as a workaround and open an issue for the IP detection.

**Comment by aserdean** [comment]
@alinbalutoiu, @rajatchopra thanks a lot for trying and testing this PR.
I addressed all of the comments and updated the commit message, can you please take a final look?
Also thanks a lot @ionutbalutoiu for hosting the new installer until we get a signed version.

File: contrib/roles/windows/kubernetes/vars/windows.yml [hunk 11]
```diff
@@ -13,7 +13,7 @@ kubernetes_info:
   K8S_DNS_DOMAIN: "{{K8S_DNS_DOMAIN | default('cluster.local')}}"
   K8S_DNS_SERVICE_IP: "{{K8S_DNS_SERVICE_IP | default('10.0.9.10')}}"
   K8S_API_SERVICE_IP: "{{K8S_API_SERVICE_IP | default('10.0.9.1')}}"
-  infracontainername_1709: "ovnkubernetes/pause"
+  infracontainername: "ovnkubernetes/pause"
```
Comment (rajatchopra) [inline contrib/roles/windows/kubernetes/vars/windows.yml]: will this eventually mean docker.io/ovnkubernetes/pause?
Let us discuss the final place for the community repository.

**Comment by rajatchopra** [comment]
/lgtm
tested ok.
